src/visor/blockdb/unspent.go

```diff
@@ -0,0 +1,248 @@
+func (up *UnspentPool) Get(h cipher.SHA256) (coin.UxOut, bool, error) {
+	if v := up.pool.Get([]byte(h.Hex())); v != nil {
+		var ux coin.UxOut
+		if err := encoder.DeserializeRaw(v, &ux); err != nil {
+			return coin.UxOut{}, false, fmt.Errorf("get uxout from pool failed: %v", err)
+		}
+		return ux, true, nil
+	}
+
+	return coin.UxOut{}, false, nil
+}
+
+// GetAll returns Pool as an array. Note: they are not in any particular order.
+func (up *UnspentPool) GetAll() (ua coin.UxArray, err error) {
+	arr := make(coin.UxArray, 0, up.pool.Len())
+	if err := up.pool.ForEach(func(k, v []byte) error {
+		var uxout coin.UxOut
+		if err := encoder.DeserializeRaw(v, &uxout); err != nil {
+			return err
+		}
+
+		arr = append(arr, uxout)
+		return nil
+	}); err != nil {
+		return coin.UxArray{}, err
+	}
+
+	return arr, nil
+}
+
+// delete removes an unsepnt from the pool by hash
+func (up *UnspentPool) delete(h cipher.SHA256) error {
+	if v := up.pool.Get([]byte(h.Hex())); v != nil {
+		var ux coin.UxOut
+		if err := encoder.DeserializeRaw(v, &ux); err != nil {
+			return err
+		}
+
+		hash, err := up.getXorHash()
+		if err != nil {
+			return err
+		}
+
+		hash = hash.Xor(ux.SnapshotHash())
+		if err := up.setXorHash(hash); err != nil {
+			return err
+		}
+
+		// delete unspent hash from pool
+		if err := up.pool.Delete([]byte(h.Hex())); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// Delete delete unspent of given hashes
+func (up *UnspentPool) Delete(hashes []cipher.SHA256) error {
+	for _, hash := range hashes {
+		if err := up.delete(hash); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// Len returns the pool size
+func (up *UnspentPool) Len() uint64 {
+	return uint64(up.pool.Len())
+}
+
+// Collides checks for hash collisions with existing hashes
+func (up *UnspentPool) Collides(hashes []cipher.SHA256) bool {
+	for i := range hashes {
+		if up.Contains(hashes[i]) {
+			return true
+		}
+	}
+	return false
+}
+
+// Contains check if the hash of uxout does exist in the pool
+func (up *UnspentPool) Contains(h cipher.SHA256) bool {
+	v := up.pool.Get([]byte(h.Hex()))
+	return v != nil
+}
+
+// GetUnspentsOfAddr returns all unspent outputs of given address
+func (up *UnspentPool) GetUnspentsOfAddr(addr cipher.Address) (coin.UxArray, error) {
+	uxs := coin.UxArray{}
+	if err := up.pool.ForEach(func(k, v []byte) error {
+		var ux coin.UxOut
+		if err := encoder.DeserializeRaw(v, &ux); err != nil {
+			return err
+		}
+
+		if ux.Body.Address == addr {
+			uxs = append(uxs, ux)
+		}
+		return nil
+	}); err != nil {
+		return coin.UxArray{}, fmt.Errorf("get unsepnts of address %v failed: %v",
+			addr.String(), err)
+	}
+
+	return uxs, nil
+}
+
+// GetUnspentsOfAddrs returns unspent outputs map of given addresses,
+// the address as return map key, unspent outputs as value.
+func (up *UnspentPool) GetUnspentsOfAddrs(addrs []cipher.Address) (coin.AddressUxOuts, error) {
+	addrm := make(map[cipher.Address]byte, len(addrs))
```

Comment:
`map[Type]struct{}` is idiomatic for lookup tables,
the line below would be `addrm[a] = struct{}{}` instead of `addrm[a] = byte(1)`.  empty struct uses no memory